lib/cretonne/src/ir/funcname.rs

```diff
@@ -14,65 +13,30 @@ use std::ascii::AsciiExt;
 pub struct FunctionName(String);
 
 impl FunctionName {
-    /// Create new function name equal to `s`.
-    pub fn new<S: Into<String>>(s: S) -> FunctionName {
+    /// Creates a new function name from a string.
+    ///
+    /// Caller should validate that the string contains only
+    /// ASCII alphanumerical characters and `_`.
+    pub fn from_string(s: &str) -> FunctionName {
         FunctionName(s.into())
     }
 }
 
-fn is_id_start(c: char) -> bool {
-    c.is_ascii() && (c == '_' || c.is_alphabetic())
-}
-
-fn is_id_continue(c: char) -> bool {
-    c.is_ascii() && (c == '_' || c.is_alphanumeric())
-}
-
-// The function name may need quotes if it doesn't parse as an identifier.
-fn needs_quotes(name: &str) -> bool {
-    let mut iter = name.chars();
-    if let Some(ch) = iter.next() {
-        !is_id_start(ch) || !iter.all(is_id_continue)
-    } else {
-        // A blank function name needs quotes.
-        true
-    }
-}
-
 impl fmt::Display for FunctionName {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        if needs_quotes(&self.0) {
-            f.write_char('"')?;
-            for c in self.0.chars().flat_map(char::escape_default) {
-                f.write_char(c)?;
-            }
-            f.write_char('"')
-        } else {
-            f.write_str(&self.0)
-        }
+        f.write_char('%')?;
+        f.write_str(&self.0)
```

Comment:
Perhaps easier: `write!(f, "%{}", self.0)`.

I would like for the `Display` implementation to still detect if the function name is pure ASCII so it can be printed with the `%` syntax. If it can't, it should be printed in hexadecimal instead with `#xxxx`. The legal characters in the `%xx` notating is ASCII alphanumerical characters and `_`.